File: Source/C#/DevnologyFitnesseDojo/DevnologyFitnesseDojo/Domain/BookService.cs

namespace DevnologyFitnesseDojo.Domain
{
    public class BookService
    {
        private readonly Inventory inventory = Inventory.Instance;

        public void BuyBooks(Order order)
        {
            int totalPriceBeforeDiscount = order.TotalPrice;

            int totalPriceAfterDiscount = totalPriceBeforeDiscount - CalculateDiscount(order);

            Customer customer = order.Customer;

            if (customer.Money >= totalPriceAfterDiscount)
            {
                customer.Money = totalPriceAfterDiscount;
            }

            // Then deduct the inventory
            foreach (Book book in order.Books)
            {
                int amount = order.GetAmount(book);

                inventory.DeductInventory(book, amount);
            }
        }

        private int CalculateDiscount(Order order)
        {
            if (!inventory.IsPromoOrder(order))
            {
                return 0;
            }

            PromoPackage promoPackage = inventory.GetPromoPackage(order);

            return promoPackage.caclulateDiscount();
        }
    }
}

File: Source/Java/nl/devnology/domain/book/BookService.java
package nl.devnology.domain.book;

import nl.devnology.domain.customer.Customer;
import nl.devnology.domain.order.Order;

/**
 * @author Erik Pragt
 */
public class BookService {
    private Inventory inventory = Inventory.get();

    public void buyBooks(Order order) {
        Integer totalPriceBeforeDiscount = order.getTotalPrice();

        Integer totalPriceAfterDiscount = totalPriceBeforeDiscount - calculateDiscount(order);

        Customer customer = order.getCustomer();

        if (customer.getMoney() >= totalPriceAfterDiscount) {
            customer.deductMoney(totalPriceAfterDiscount);
        }

        // Then deduct the inventory
        for (Book book : order.getBooks()) {
            Integer amount = order.getAmount(book);

            Inventory.get().deductInventory(book, amount);
        }
    }

    private Integer calculateDiscount(Order order) {
        if (!inventory.isPromoOrder(order)) {
            return 0;
        }

        PromoPackage promoPackage = inventory.getPromoPackage(order);

        return promoPackage.caclulateDiscount();
    }
}
